File: my_cpp_pkg/src/activity_02/number_counter.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/int64.hpp"

class NumberCounter : public rclcpp::Node
{
public:
    NumberCounter() : Node("number_counter")
    {
        m_publisher = this->create_publisher<example_interfaces::msg::Int64>("number_count", 10);
        m_subscription = this->create_subscription<example_interfaces::msg::Int64>(
            "number", 10, std::bind(&NumberCounter::CallbackNumber, this, std::placeholders::_1));
        RCLCPP_INFO(this->get_logger(), "Number counter node has been started");
        RCLCPP_INFO(this->get_logger(), "Counter value: '%ld'", m_count);
    }

private:
    int64_t m_count = 0;
    rclcpp::Subscription<example_interfaces::msg::Int64>::SharedPtr m_subscription;
    rclcpp::Publisher<example_interfaces::msg::Int64>::SharedPtr m_publisher;

    void CallbackNumber(const example_interfaces::msg::Int64::SharedPtr msg)
    {
        m_count += msg->data;
        RCLCPP_INFO(this->get_logger(), "Received number: '%ld'", msg->data);
        RCLCPP_INFO(this->get_logger(), "Counter value: '%ld'", m_count);
        PublishCount(m_count);
    }

    void PublishCount(const int64_t count)
    {
        auto message = example_interfaces::msg::Int64();
        message.data = count;
        m_publisher->publish(message);
    }
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<NumberCounter>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: my_py_pkg/my_py_pkg/activity_02/number_counter.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from example_interfaces.msg import Int64

class NumberCounterNode(Node):
    def __init__(self):
        super().__init__("number_counter")
        self._counter = 0
        self._number_count_publisher = self.create_publisher(
            Int64, "number_count", 10)
        self._number_subscriber = self.create_subscription(
            Int64, "number", self.callback_number, 10)
        self.get_logger().info("Number Counter has been started.")

    def callback_number(self, msg: Int64):
        self._counter += msg.data
        new_msg = Int64()
        new_msg.data = self._counter
        self._number_count_publisher.publish(new_msg)


def main(args=None):
    rclpy.init(args=args)
    node = NumberCounterNode()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == "__main__":
    main()
